charms/istio-pilot/src/resources_handler.py

```diff
@@ -0,0 +1,118 @@
+
+
+class ResourceHandler:
+    def __init__(self, app_name, model_name):
+        """A Lightkube API interface.
+
+        Args:
+           - app_name: name of the application
+           - model_name: name of the Juju model this charm is deployed to
+        """
+
+        self.app_name = app_name
+        self.model_name = model_name
+
+        self.log = logging.getLogger(__name__)
+
+        # Every lightkube API call will use the model name as the namespace by default
+        self.lightkube_client = Client(namespace=self.model_name, field_manager="lightkube")
+
+        self.env = Environment(loader=FileSystemLoader('src'))
+
+    def delete_object(
+        self, obj, namespace=None, ignore_not_found=False, ignore_unauthorized=False
+    ):
+        try:
+            self.lightkube_client.delete(type(obj), obj.metadata.name, namespace=namespace)
+        except ApiError as err:
+            self.log.exception("ApiError encountered while attempting to delete resource.")
+            if err.status.message is not None:
+                if "not found" in err.status.message and ignore_not_found:
+                    self.log.error(f"Ignoring not found error:\n{err.status.message}")
+                elif "(Unauthorized)" in err.status.message and ignore_unauthorized:
+                    # Ignore error from https://bugs.launchpad.net/juju/+bug/1941655
+                    self.log.error(f"Ignoring unauthorized error:\n{err.status.message}")
+                else:
+                    self.log.error(err.status.message)
+                    raise
+            else:
+                raise
+
+    def delete_existing_resource_objects(
+        self,
+        resource,
+        namespace=None,
+        ignore_not_found=False,
+        ignore_unauthorized=False,
+        labels={},
+    ):
+        for obj in self.lightkube_client.list(
+            resource,
+            labels={"app.juju.is/created-by": f"{self.app_name}"}.update(labels),
+            namespace=namespace,
+        ):
+            self.delete_object(
+                obj,
+                namespace=namespace,
+                ignore_not_found=ignore_not_found,
+                ignore_unauthorized=ignore_unauthorized,
+            )
+
+    def apply_manifest(self, manifest, namespace=None):
+        for obj in codecs.load_all_yaml(manifest):
+            self.lightkube_client.apply(obj, namespace=namespace)
+
+    def delete_manifest(
+        self, manifest, namespace=None, ignore_not_found=False, ignore_unauthorized=False
+    ):
+        for obj in codecs.load_all_yaml(manifest):
+            self.delete_object(
+                obj,
+                namespace=namespace,
+                ignore_not_found=ignore_not_found,
+                ignore_unauthorized=ignore_unauthorized,
+            )
+
+    def create_ns_resource(self, filename: str, context: dict = None) -> GenericNamespacedResource:
```

Comment:
I think after a refactor to provide default values for `context`, the `context` in the function args is unused.  Is the intent to use the below defaults always, or just when the function is called without some of the fields?

Reply:
Unrelated to above, when reading through code that invokes `create_ns_resource` I interpreted it as creating the **resource** (in k8s), not the Lightkube resource object that represents the resource.  I'm not sure how to improve this, but if you can think of something to help that would be great.  I realise this name probably derives from Lightkube calling their own functions `create_namespaced_resource`, so maybe it makes sense to leave it, but still something to consider.

Also, if you're really making a wrapper around the lightkube function, I'd consider either calling it the same name (`create_namespaced_resource`) or something meaningfully different (`create_lightkube_generic_resource_class`).  Shortening the `namespaced` to `ns` but leaving it the same felt like an awkward middle ground.  But that's just one person's opinion :)

Reply:
Regarding your first comment, yes `context` is an unused arg. I recently changed this bit, removed the arg from the docstrings, but forgot the args list in the method definition. Will change that, good catch.

Regarding your second comment, yeh, I also felt awkward naming this method. I didn't want it to be named exactly the same as the Lightkube method to avoid confusion. I'd rather have something meaningfully different, in fact `get_generic_resource_class` or a shorter `generic_resource_class` could be more accurate as the method actually returns the resource class. What do you think?

Reply:
Changed it from `create_ns_resource` to `generate_generic_resource_class`, let me know if that is better.